libraries/chain/include/eosio/chain/fork_database.hpp

```diff
@@ -157,16 +161,18 @@ namespace eosio::chain {
       // see fork_database_t::fetch_branch(forkdb->head()->id())
       block_branch_t fetch_branch_from_head() const;
 
+      void reset_root(const block_handle::block_handle_variant_t& v);
```

Comment:
I really dislike that this exposes `block_handle::block_handle_variant_t`. I would prefer that type to not escape controller.cpp

Reply:
In my mind this is even worse. I would say we either go all the way and provide every method that takes a variant or we continue to use apply/etc in controller. I say we keep with using apply in controller.

Reply:
Hum, I don't think it is worse.

I get that `block_handle` is its own class providing a unified  view of a `block_state` and that we don't want other things to depend on its internals.

However, it seems that there are some cases where we need the new `block_state_variant_t` type, like when we want to reset the root of our `external` fork_database which contains two fork_db pointers. I don't see why it is bad to define this type?

My intent in doing this was to try to help with the transition case where we may have two simultaneous fork_dbs. It may not be the best solution for your actual needs. If you know at this point what the interface should be, I'm happy to update this PR. 

Did you mean in your comment that I should replace all instances of `block_state_variant_t` with `std::variant<block_state_legacy_ptr, block_state_ptr>`?

Reply:
> Did you mean in your comment that I should replace all instances of block_state_variant_t with std::variant<block_state_legacy_ptr, block_state_ptr>?

No. We use apply for this type of thing elsewhere, just seem to me to be consisted to keep using apply. In my PR the creation of the `new_head` is in controller.

Reply:
You can leave as is, it will go away with my PR anyway.

Reply:
Sounds good, I'll leave as-is, feel free to update as you see fit. 
I don't think we can use controller's `apply` always because what we want to do may depend of `in_use` inside fork_db.